socketcan/libcanard/test/test_io.cpp

```diff
@@ -85,28 +86,71 @@ TEST_CASE("IO-FD")
 
     char buf[255]{};
     fr = {};
-    REQUIRE(1 == socketcanPop(sb, &fr, sizeof(buf), buf, 1000));
+    REQUIRE(1 == socketcanPop(sb, &fr, sizeof(buf), buf, 1000, nullptr));
     REQUIRE(fr.timestamp_usec > 0);
     REQUIRE(fr.extended_can_id == 0x1234U);
     REQUIRE(fr.payload_size == 13);
     REQUIRE(0 == std::memcmp(fr.payload, "Hello world!", 13));
     auto old_ts = fr.timestamp_usec;
 
     fr = {};
-    REQUIRE(1 == socketcanPop(sa, &fr, sizeof(buf), buf, 1000));
+    REQUIRE(0 == socketcanPop(sa, &fr, sizeof(buf), buf, 1000, nullptr));  // Loopback frame.
+    REQUIRE(1 == socketcanPop(sa, &fr, sizeof(buf), buf, 1000, nullptr));  // Received actual frame.
     REQUIRE(fr.timestamp_usec > 0);
     REQUIRE(fr.timestamp_usec >= old_ts);
     REQUIRE(fr.extended_can_id == 0x4321U);
     REQUIRE(fr.payload_size == 10);
     REQUIRE(0 == std::memcmp(fr.payload, "0123456789", 10));
 
-    REQUIRE(0 == socketcanPop(sa, &fr, sizeof(buf), buf, 0));
-    REQUIRE(0 == socketcanPop(sa, &fr, sizeof(buf), buf, 1000));
-    REQUIRE(-EINVAL == socketcanPop(sa, &fr, 0, nullptr, 1000));
-    REQUIRE(-EINVAL == socketcanPop(sa, nullptr, sizeof(buf), buf, 1000));
+    REQUIRE(0 == socketcanPop(sa, &fr, sizeof(buf), buf, 0, nullptr));
+    REQUIRE(0 == socketcanPop(sa, &fr, sizeof(buf), buf, 1000, nullptr));
+    REQUIRE(-EINVAL == socketcanPop(sa, &fr, 0, nullptr, 1000, nullptr));
+    REQUIRE(-EINVAL == socketcanPop(sa, nullptr, sizeof(buf), buf, 1000, nullptr));
 
     REQUIRE(-EINVAL == socketcanPush(sa, nullptr, 1'000'000));
 
     ::close(sa);
     ::close(sb);
 }
+
+TEST_CASE("IO-FD-Loopback")
+{
+    const char* const iface_name = "vcan0";
+
+    const auto sa = socketcanOpen(iface_name, true);
+    const auto sb = socketcanOpen(iface_name, true);
+    REQUIRE(sa >= 0);
+    REQUIRE(sb >= 0);
+
+    CanardFrame fr{};
+    fr.extended_can_id = 0x1234U;
+    fr.payload_size    = 13;
+    fr.payload         = "Hello World!";
+    REQUIRE(1 == socketcanPush(sa, &fr, 1'000'000));  // Send frame on sa.
+
+    bool loopback = true;
+    char buf[255]{};
+    fr = {};
+    REQUIRE(1 == socketcanPop(sb, &fr, sizeof(buf), buf, 1000, &loopback));  // Receive actual frame on sb.
+    REQUIRE(loopback == false);
+    REQUIRE(fr.timestamp_usec > 0);
+    REQUIRE(fr.extended_can_id == 0x1234U);
+    REQUIRE(fr.payload_size == 13);
+    REQUIRE(0 == std::memcmp(fr.payload, "Hello World!", 13));
+    auto old_ts = fr.timestamp_usec;
+
+    fr = {};
+    REQUIRE(1 == socketcanPop(sa, &fr, sizeof(buf), buf, 1000, &loopback));  // Receive loopback frame on sa.
+    REQUIRE(loopback == true);
+    REQUIRE(fr.timestamp_usec > 0);
+    REQUIRE(fr.timestamp_usec >= old_ts); // TODO: time relation still sensible? Appears to be equal...
```

Comment:
This seems right, yes. They may appear equal depending on the resolution of the timestamping clock.